### PR TITLE
✨  (INT-2183): Customer Endpoint Reference Service Implementation

### PR DESCRIPTION
**Pull request by fahimmahmoodmir**
- Implementation of the Custom Endpoint Reference Service

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,5 @@
+root = true
+
+[*.js]
```
Comment (cscharun) [inline .editorconfig]: If this file is necessary, please update it for *.ts

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+.pnp.*
+.npm
+.yarn
+.yarn-integrity
+*.tgz
+
+# IDEs
```
Comment (cscharun) [inline .gitignore:22]: You can check https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore to be more precise with jetbrain products.

File: .prettierrc [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "singleQuote": false,
+  "trailingComma": "none",
```
Comment (cscharun) [inline .prettierrc:3]: Was this by purpose?
Why not having the es5 approach with trailing commas for objects and arrays?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,17 @@
+FROM node:16-alpine as tpb_custom_endpoint_service
```
Comment (cscharun) [inline Dockerfile]: If we don't run the container in development, only dist should be copied into the container.

**Review by cscharun — COMMENTED** [review]
Running out of time. Will continue my review later.

File: .prettierrc (hunk 3 above)
Comment (fahimmahmoodmir) [reply .prettierrc:3]: Yep to comply with our own styles of coding in other projects

File: .editorconfig (hunk 1 above)
Comment (fahimmahmoodmir) [reply .editorconfig]: Yep for vscode

File: README.md [hunk 5]
```diff
@@ -1,49 +1,266 @@
...
+and [src/controllers/responseBodies](src/controllers/responseBodies) for Request and Response bodies respectively.
 
-The OpenAPI specification can be found under /src/spec
-x-middlewares fields are not considered part of the specification and can be ignored.
+## Service Specification Concepts
 
-The specification describes a REST API that can support different versions of multiple bots.
-The URL for a bot instance looks like the following example
+> Note: Detailed concepts of Service Implementation can be found in the official documentation of
+> Custom Endpoint [here][cbserimpl]
```
Comment (cscharun) [inline README.md:19]: Add link here

File: README.md [hunk 6]
```diff
@@ -1,49 +1,266 @@
...
-Third-Party Bots will authenticate its requests to the Custom Endpoint API by sending a signed JWT generated using [Private Conversational Cloud Applications](https://developers.liveperson.com/conversational-cloud-applications-what-is-a-conversational-cloud-application.html).
+
+Third-Party Bots will authenticate its requests to the Custom Endpoint API by sending a signed JWT generated
+using private [conversational cloud applications][ccapp].
 
 ### Bots
-A bot is identified by its uuid or a similar alphanumeric identifier.
-A service implementation can restrict access to bots bases on the accoundId provided as the 'aud' field on the authentication WebToken
+
+A bot is identified by its UUID or a similar alphanumeric identifier.A service implementation can restrict
```
Comment (cscharun) [inline README.md]:
```suggestion
A bot is identified by its UUID or a similar alphanumeric identifier.A service implementation can restrict
```
```suggestion
A bot is identified by its UUID or a similar alphanumeric identifier. A service implementation can restrict
```

File: README.md [hunk 7]
```diff
@@ -1,49 +1,266 @@
...
+access to bots bases on the `accountId` provided as the 'aud' field on the authentication WebToken
```
Comment (cscharun) [inline README.md]:
```suggestion
access to bots bases on the `accountId` provided as the 'aud' field on the authentication WebToken
```
```suggestion
access to bots based on the `accountId` provided as the 'aud' field on the authentication WebToken
```

Comment (cscharun) [reply README.md]: Wouldn't it be better to say something like:

'A bot is identified by its UUID or a similar alphanumeric identifier. The supplied JWT can be used to restrict access to bots as it contains account specific properties, e.g. the `accountId` provided as the 'aud' field.'

File: README.md [hunk 8]
```diff
@@ -1,49 +1,266 @@
...
+## Features
+
+- Contains implementation of `v1` of the Custom Endpoint API specification
+- Contains supported Bot response examples.
+- Contains the OpenAPI specification (found under `/src/docs/`)
+- Contains local caching for the calls to the LivePerson APIs
+
+### Bots
+
+in this service implementation we have configured some bots model manually that can be used in Vendor
```
Comment (cscharun) [inline README.md]:
```suggestion
in this service implementation we have configured some example bots manually that can be used in Vendor
```

File: README.md [hunk 9]
```diff
@@ -1,49 +1,266 @@
...
+
+- `5809777a-e548-4bd2-bc5f-a1003c132a28` - Bot with online state
+- `36d5de7f-6c83-41ba-836d-28ca1062c1ef` - Bot with offline state
+- `42773157-adcf-4ec1-a2ae-9b935d80cb81` - Bot with error state
+- `213aa6c6-f6d9-11ec-b939-0242ac120002` - Bot with draft state
+- `bfb807d6-f6d8-11ec-b939-0242ac120002` - Bot with maintenance state
+
+### Supported Customer Utterances
+
+In this service implementation we have configured some Bot Responses that are supported by Third-Party Bots.
```
Comment (cscharun) [inline README.md]:
```suggestion
In this service implementation we have configured some `Bot Responses` that are supported by Third-Party Bots.
```

File: README.md [hunk 10]
```diff
@@ -1,49 +1,266 @@
...
+These response are defined in [src/utils/Response.ts](src/utils/Response.ts) On sending customer utterances (as shown in below table)
```
Comment (cscharun) [inline README.md]:
```suggestion
These responses are defined in [src/utils/Response.ts](src/utils/Response.ts). On sending customer utterances (as shown in below table)
```

File: src/controllers/BotController.ts [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+          data as IRichContentEventData
+        ) as PostConversationEventsResponseBody;
+        break;
+      case EventMessageType.TEXT:
+        return this.botService.fetchConversationTextEvents(
+          botId,
+          data as ITextEventData
+        );
+        break;
+      case EventMessageType.START:
```
Comment (cscharun) [inline src/controllers/BotController.ts:119]: Welcome Event is also cloaked as START event

File: src/index.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    format: winston.format.combine(
+      winston.format.json(),
+      winston.format.colorize()
+    )
+  })
+);
+
+useExpressServer(app, {
+  authorizationChecker: async (action: Action) => {
+    const token = action.request.headers["authorization"].split(" ")[1];
```
Comment (cscharun) [inline src/index.ts]:
```suggestion
    const authHeader = action.request.headers.authorization;

    if (authHeader) {
        const token = authHeader.split(' ')[1];
        
        ...
 }
```

File: src/index.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    transports: [new winston.transports.Console()],
+    format: winston.format.combine(
+      winston.format.json(),
+      winston.format.colorize()
+    )
+  })
+);
+
+useExpressServer(app, {
+  authorizationChecker: async (action: Action) => {
```
Comment (cscharun) [inline src/index.ts]: Can you import the authorization checker from a module?

File: src/models/bots/responses/IStructuredContentResponse.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import { IResponseBase } from "./IResponseBase";
+import { IResponseType } from "./IResponseTypes";
+
+export type StructureContentData = {
```
Comment (cscharun) [inline src/models/bots/responses/IStructuredContentResponse.ts]:
```suggestion
export type StructuredContentData = {
```

File: src/models/bots/IBot.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+export type IBot = {
+  id: string;
+  getId(): string;
+  getVersion(): string;
+  getState(): string;
```
Comment (cscharun) [inline src/models/bots/IBot.ts]: State can be of State enum type. You already did this enum.

**Review by cscharun — APPROVED** [review]
Good work. Only have some smaller recommendations for you.

File: .gitignore (hunk 2 above)
Comment (fahimmahmoodmir) [reply .gitignore:22]: sure will add those in ignore files as well

File: Dockerfile (hunk 4 above)
Comment (fahimmahmoodmir) [reply Dockerfile]: sure

File: README.md (hunk 5 above)
Comment (fahimmahmoodmir) [reply README.md:19]: link for? its already linked ^^

File: src/models/bots/IBot.ts (hunk 15 above)
Comment (fahimmahmoodmir) [reply src/models/bots/IBot.ts]: indeed it can be

File: src/index.ts (hunk 13 above)
Comment (fahimmahmoodmir) [reply src/index.ts]: sure

File: src/controllers/BotController.ts (hunk 11 above)
Comment (fahimmahmoodmir) [reply src/controllers/BotController.ts:119]: What you mean i don't understand?

**Pull request merged**